### PR TITLE
✨ feat: Remove owned VPCEs when deleting VPC.

### LINKED ISSUE #5192: Remove all `owned` VPCEs during VPC deletion
/kind feature

**Describe the solution you'd like**
CAPA should remove all `owned` VPCEs that are bound to VPC, during VPC deletion.

In case of EKS ManagedControlPlane deletion, when `owned` VPC gets deleted [this code](https://github.com/kubernetes-sigs/cluster-api-provider-aws/blob/main/pkg/cloud/services/network/vpc.go#L280-L286) is a no-op [due to no s3 for ManagedControlPlane](https://github.com/kubernetes-sigs/cluster-api-provider-aws/blob/main/pkg/cloud/scope/managedcontrolplane.go#L334-L338) 

**Anything else you would like to add:**
ref: October 28, 2024 Meeting Notes

**Environment:**

- Cluster-api-provider-aws version: v2.6.1
- Kubernetes version: (use `kubectl version`): v1.31.0

### PR DESCRIPTION
**Pull request by s3rj1k**
/kind feature

Fixes #5192 

**Special notes for your reviewer**:

**Checklist**:

- [x] squashed commits
- [ ] includes documentation
- [x] includes [emojis](https://github.com/kubernetes-sigs/kubebuilder-release-tools?tab=readme-ov-file#kubebuilder-project-versioning)
- [ ] adds unit tests
- [ ] adds or updates e2e tests

```release-note
NONE
```

### CODE REVIEW

**Comment by s3rj1k** [comment]
/test pull-cluster-api-provider-aws-test

Hmm, running locally tests passed
```
KUBEBUILDER_ASSETS="/root/.local/share/kubebuilder-envtest/k8s/1.30.2-linux-arm64" /usr/bin/go test -timeout 30s -run ^TestAWSClusterReconcilerIntegrationTests$ sigs.k8s.io/cluster-api-provider-aws/v2/controllers
ok  	sigs.k8s.io/cluster-api-provider-aws/v2/controllers	19.045s
```

**Comment by s3rj1k** [comment]
Tests are green 🎉

**Comment by s3rj1k** [comment]
Tests are green 🎉 including `eks`

**Comment by s3rj1k** [comment]
@richardcase @dlipovetsky Please take a look

File: pkg/cloud/services/network/vpc.go [hunk 1]
```diff
@@ -276,17 +276,8 @@ func (s *Service) deleteVPCEndpoints() error {
...
-	if s.scope.Bucket() != nil {
-		services.Insert(fmt.Sprintf("com.amazonaws.%s.s3", s.scope.Region()))
-	}
-	if services.Len() == 0 {
-		return nil
-	}
-
 	// Get all existing endpoints.
-	endpoints, err := s.describeVPCEndpoints()
+	endpoints, err := s.describeVPCEndpoints(filter.EC2.Cluster(s.scope.Name()))
```
Comment (richardcase) [inline pkg/cloud/services/network/vpc.go:280]: I would say that we want to make sure its `owned` and not `shared`, mainly to be consistent with other places in the code.

```suggestion
	endpoints, err := s.describeVPCEndpoints(filter.EC2.ClusterOwned(s.scope.Name()))
```

File: test/e2e/shared/suite.go [hunk 2]
```diff
@@ -145,7 +145,7 @@ func Node1BeforeSuite(e2eCtx *E2EContext) []byte {
 				success = false
 			}
 			return success
-		}, 10*time.Minute, 5*time.Second).Should(BeTrue(), "Should've eventually succeeded creating an AWS CloudFormation stack")
+		}, 30*time.Minute, 15*time.Second).Should(BeTrue(), "Should've eventually succeeded creating an AWS CloudFormation stack")
```
Comment (richardcase) [inline test/e2e/shared/suite.go:148]: Was there a reason you changed this from 10 minutes to 30 mins?

Comment (s3rj1k) [reply test/e2e/shared/suite.go:148]: Yea, tests where failing on timeouts few times, to reduce re-triggers

Comment (s3rj1k) [reply test/e2e/shared/suite.go]: Seems to still timeout a lot...

**Comment by richardcase** [comment]
@s3rj1k - if the e2e doesn't fail after a few more retests then we can dig into why its failing.

**Comment by s3rj1k** [comment]
@richardcase that `pull-cluster-api-provider-aws-e2e-eks` is supper-flaky

On a side note, I managed to test this PR on actual EKS cluster, so generally it should work.

**Comment by richardcase** [comment]
A flake with the cloudformation creation, so:

/test pull-cluster-api-provider-aws-e2e

**Comment by s3rj1k** [comment]
due to "Should've eventually succeeded creating an AWS CloudFormation stack"
/test pull-cluster-api-provider-aws-e2e

**Comment by s3rj1k** [comment]
due to "Timed out after 2700.003s. Should've eventually succeeded creating an AWS CloudFormation stack"

/test pull-cluster-api-provider-aws-e2e-eks

**Comment by s3rj1k** [comment]
One out of 1 tests failed with "Timed out after 1800.000s.
Eventually failed waiting for EKS addon "coredns" to have status ["ACTIVE" "DEGRADED"] for EKS cluster"

/test pull-cluster-api-provider-aws-e2e-eks

**Comment by richardcase** [comment]
@s3rj1k - they passed :)

**Comment by richardcase** [comment]
Capturing that the e2e passed for both EKS and non-EKS:

- https://prow.k8s.io/view/gs/kubernetes-ci-logs/pr-logs/pull/kubernetes-sigs_cluster-api-provider-aws/5193/pull-cluster-api-provider-aws-e2e/1855912691483283456
- https://prow.k8s.io/view/gs/kubernetes-ci-logs/pr-logs/pull/kubernetes-sigs_cluster-api-provider-aws/5193/pull-cluster-api-provider-aws-e2e-eks/1856073958562992128

**Comment by richardcase** [comment]
From my side:

/approve

**Comment by richardcase** [comment]
For lgtm: 

/assign @AndiDog 
/assign @nrb 
/assign @dlipovetsky 
/assign @damdo

**Comment by s3rj1k** [comment]
@richardcase Brute-forced e2e into passing :)

**Review by damdo — COMMENTED** [review]
Nice fix.

/lgtm

**Pull request merged**